### PR TITLE
Low level native contract binding subscription support

### PR DESCRIPTION
**Pull request by awishformore**
I was looking at getting the new RPC client subscription functionality into the native Go contract bindings, mostly because I am personally eager to start working on some Go contract daemons.

The PR is not complete yet, so feel free to give me tons and tons of feedback (the more the better!). I felt this is more constructive than bothering Peter privately on Gitter every day.

I also tried lots of ways to go around passing on errors on to the consumer and how to notify it of a closed subscription; in the end, I'm quite sure that closing the channel on the producer side (as is idiomatic in Go) is the cleanest way, so I added that to the RPC client.

Missing:
- implementation of the simulated backend functionality
- including the subscription functionality in the template for high level contract binding generation

Questions:
- instead of returning the error on unsubscribe, does it merrit a separate error function?
- where should I put the `SubscribeResult` entity? Is it fine where it is now?
- I'm not sure context is fully used in the subscribe functionality, any hints?
- any ideas about how to implement context with the unsubscribe?

Cheers.

P.S. I'll squash when I have something decent.

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Tue Aug 9 18:30:06 UTC 2016_

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: e326e3840476122cd27dfa6531208fb339013b4e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 29869386df95c595e77051978378b6c2c2fff96e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 4de106e99de890d9e02cec4bf57c28cc62f12cb9
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by fjl** [comment]
Hehe, looks like there is a lot of overlap with [the work that I do right now](http://twurst.com/articles/geth-1.5-api.html). I am currently working on implementing the content of that post and the API described there will be available in the 1.5.0 go-ethereum release. 

I definitely want to see support for ABI-decoded contract events happen in 1.5 as well. If you want to help, let's have a call or chat to coordinate.

> I also tried lots of ways to go around passing on errors on to the consumer and how to notify it of a > closed subscription; in the end, I'm quite sure that closing the channel on the producer side (as is > idiomatic in Go) is the cleanest way, so I added that to the RPC client.

Subscriptions have two methods, `Unsubscribe` and `Err`. The `Err` method gives you a channel
that will receive a value if there is an error with the subscription. I don't want to close the user-provided channel because that limits how you can use it on the caller side. With the current model,
the channel that you subscribe can be used for other things in addition to the subscription, and you can create multiple subscriptions on the same channel (e.g. 'logs' and 'pending logs').

> where should I put the SubscribeResult entity? Is it fine where it is now?

`SubscribeResult` is basically `vm.Log` and that's what is used in the API proposal.
We are considering to move Log to the core/types package.

> I'm not sure context is fully used in the subscribe functionality, any hints?

`EthSubscribe` takes a context parameter on the develop branch since yesterday ;)
It is used to cancel the subscribe request.

> Instead of returning the error on unsubscribe, does it merrit a separate error function?

Returning the error is fine.

> any ideas about how to implement context with the unsubscribe?

No. Unsubscribe has a 5s timeout. If you don't want to block that long you can always `go sub.Unsubscribe()`.

**Comment by awishformore** [comment]
After discussing with @fjl , it turns out the best way to go about this is to use the upcoming Ethereum stable API interfaces for the bound contracts, replacing the direct use of the RPC client and allowing future use of the contract bindings directly with an embedded full Ethereum node, light Ethereum node or over RPC through another wrapper.

@fjl said he would create a PR for just the interfaces, in which case I could start building directly against develop on this repository; otherwise, I will start work based on his branch.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: e326e3840476122cd27dfa6531208fb339013b4e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 29869386df95c595e77051978378b6c2c2fff96e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 4de106e99de890d9e02cec4bf57c28cc62f12cb9
  * Commits must be prefixed with the package(s) they modify

 * Commit: 4bc47fb2b31bc3b02e82fb7188ec337f2675ab52
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: e326e3840476122cd27dfa6531208fb339013b4e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 29869386df95c595e77051978378b6c2c2fff96e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 4de106e99de890d9e02cec4bf57c28cc62f12cb9
  * Commits must be prefixed with the package(s) they modify

 * Commit: 4bc47fb2b31bc3b02e82fb7188ec337f2675ab52
  * Commits must be prefixed with the package(s) they modify

 * Commit: e9d5d5e3c85b887871e28a7e7624523dac9ca9cc
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: accounts/abi/bind/base.go [hunk 1]
```diff
@@ -107,9 +120,9 @@ func (c *BoundContract) Call(opts *CallOpts, result interface{}, method string,
 	}
 	// Make sure we have a contract to operate on, and bail out otherwise
 	if (opts.Pending && atomic.LoadUint32(&c.pendingHasCode) == 0) || (!opts.Pending && atomic.LoadUint32(&c.latestHasCode) == 0) {
-		if code, err := c.caller.HasCode(opts.Context, c.address, opts.Pending); err != nil {
+		if code, err := c.backend.CodeAt(opts.Context, c.address, nil); err != nil {
```
Comment (fjl) [inline accounts/abi/bind/base.go]: If opts.Pending is true, this check should use `PendingCodeAt`.

File: accounts/abi/bind/base.go [hunk 2]
```diff
@@ -123,7 +136,12 @@ func (c *BoundContract) Call(opts *CallOpts, result interface{}, method string,
 	if err != nil {
 		return err
 	}
-	output, err := c.caller.ContractCall(opts.Context, c.address, input, opts.Pending)
+	// Create the call message we use to make the call
+	callMsg := ethereum.CallMsg{
+		To:   c.address,
+		Data: input,
+	}
+	output, err := c.backend.CallContract(opts.Context, callMsg, nil)
```
Comment (fjl) [inline accounts/abi/bind/base.go]: Same here, `PendingCallContract` should be used if opts.Pending is true.

Comment (fjl) [inline accounts/abi/bind/base.go]: @karalabe I didn't really look at this previously, the caching here looks unsafe. The contract creation can get reverted at any time, in which case the check will not be executed. The call will still fail, but not with ErrNoCode.

IMHO it would be safer to do the check after the call.

File: accounts/abi/bind/base.go [hunk 3]
```diff
@@ -204,8 +230,103 @@ func (c *BoundContract) transact(opts *TransactOpts, contract *common.Address, i
...
+	return nil
+}
+
+// Error returns the error that was encountered for the subscription on the given
+// channel. It will reset the error upon retrieval.
+func (c *BoundContract) Error(channel chan<- vm.Log) error {
+	err := c.errors[channel]
+	delete(c.errors, channel)
+	return err
+}
```
Comment (fjl) [inline accounts/abi/bind/base.go]: This is racing vs the write above.

File: rpc/client.go [hunk 4]
```diff
@@ -344,6 +344,22 @@ func (c *Client) BatchCallContext(ctx context.Context, b []BatchElem) error {
...
+// sent to the given channel. The element type of the channel must match the
+// expected type of content returned by the subscription. EthSubscribeContext
+// returns as soon as the given context expires.
+//
+// Callers should not use the same channel for multiple calls to EthSubscribe.
+// The channel is closed when the notification is unsubscribed or an error
+// occurs. The error can be retrieved via the Err method of the subscription.
+//
+// Slow subscribers will block the clients ingress path eventually.
+func (c *Client) EthSubscribeContext(ctx context.Context, channel interface{}, args ...interface{}) (*ClientSubscription, error) {
```
Comment (fjl) [inline rpc/client.go]: Please rebase on develop. EthSubscribe takes a context argument there.

File: rpc/client.go [hunk 5]
```diff
@@ -687,6 +701,7 @@ func (sub *ClientSubscription) quitWithError(err error, unsubscribeServer bool)
 			sub.err <- err
 		}
 		close(sub.quit)
+		sub.channel.Close()
```
Comment (fjl) [inline rpc/client.go]: Please revert the closing.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 6fb1523944824d15b2f88f113747af22174c9045
  * Commits must be prefixed with the package(s) they modify

 * Commit: 4812a4f49e78220f617737cda909c54075376cc5
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by karalabe** [comment]
Hmmm, seems there's a lot of refactoring going on in here that doesn't have much to do with events. I'd really suggest that we do events and refactors in separate PRs. It's hard to follow one or the other when everything seems to change.

Perhaps if there's a need to rewrite parts of the internals then we should do that first and then add the events?

**Comment by awishformore** [comment]
I will replace this by one PR to use the new stable Go API instead of the current backends, and another PR to add the subscription support.
